File: python/src/acedatacloud/resources/webextrator.py
"""WebExtrator web render & extract resources."""

from __future__ import annotations

from typing import Any, Literal


class WebExtrator:
    """Synchronous WebExtrator client."""

    def __init__(self, transport: Any) -> None:
        self._transport = transport

    def extract(
        self,
        *,
        url: str,
        expected_type: Literal["product", "article", "general"] | None = None,
        enable_llm: bool | None = None,
        wait_until: Literal["load", "domcontentloaded", "networkidle", "commit"] | None = None,
        timeout: float | None = None,
        delay: float | None = None,
        wait_for_selector: str | None = None,
        block_resources: list[str] | None = None,
        headers: dict[str, str] | None = None,
        user_agent: str | None = None,
        callback_url: str | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"url": url, **kwargs}
        if expected_type is not None:
            body["expected_type"] = expected_type
        if enable_llm is not None:
            body["enable_llm"] = enable_llm
        if wait_until is not None:
            body["wait_until"] = wait_until
        if timeout is not None:
            body["timeout"] = timeout
        if delay is not None:
            body["delay"] = delay
        if wait_for_selector is not None:
            body["wait_for_selector"] = wait_for_selector
        if block_resources is not None:
            body["block_resources"] = block_resources
        if headers is not None:
            body["headers"] = headers
        if user_agent is not None:
            body["user_agent"] = user_agent
        if callback_url is not None:
            body["callback_url"] = callback_url
        return self._transport.request("POST", "/webextrator/extract", json=body)

    def render(
        self,
        *,
        url: str,
        wait_until: Literal["load", "domcontentloaded", "networkidle", "commit"] | None = None,
        timeout: float | None = None,
        delay: float | None = None,
        wait_for_selector: str | None = None,
        block_resources: list[str] | None = None,
        headers: dict[str, str] | None = None,
        user_agent: str | None = None,
        callback_url: str | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"url": url, **kwargs}
        if wait_until is not None:
            body["wait_until"] = wait_until
        if timeout is not None:
            body["timeout"] = timeout
        if delay is not None:
            body["delay"] = delay
        if wait_for_selector is not None:
            body["wait_for_selector"] = wait_for_selector
        if block_resources is not None:
            body["block_resources"] = block_resources
        if headers is not None:
            body["headers"] = headers
        if user_agent is not None:
            body["user_agent"] = user_agent
        if callback_url is not None:
            body["callback_url"] = callback_url
        return self._transport.request("POST", "/webextrator/render", json=body)


class AsyncWebExtrator:
    """Async WebExtrator client."""

    def __init__(self, transport: Any) -> None:
        self._transport = transport

    async def extract(
        self,
        *,
        url: str,
        expected_type: Literal["product", "article", "general"] | None = None,
        enable_llm: bool | None = None,
        wait_until: Literal["load", "domcontentloaded", "networkidle", "commit"] | None = None,
        timeout: float | None = None,
        delay: float | None = None,
        wait_for_selector: str | None = None,
        block_resources: list[str] | None = None,
        headers: dict[str, str] | None = None,
        user_agent: str | None = None,
        callback_url: str | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"url": url, **kwargs}
        if expected_type is not None:
            body["expected_type"] = expected_type
        if enable_llm is not None:
            body["enable_llm"] = enable_llm
        if wait_until is not None:
            body["wait_until"] = wait_until
        if timeout is not None:
            body["timeout"] = timeout
        if delay is not None:
            body["delay"] = delay
        if wait_for_selector is not None:
            body["wait_for_selector"] = wait_for_selector
        if block_resources is not None:
            body["block_resources"] = block_resources
        if headers is not None:
            body["headers"] = headers
        if user_agent is not None:
            body["user_agent"] = user_agent
        if callback_url is not None:
            body["callback_url"] = callback_url
        return await self._transport.request("POST", "/webextrator/extract", json=body)

    async def render(
        self,
        *,
        url: str,
        wait_until: Literal["load", "domcontentloaded", "networkidle", "commit"] | None = None,
        timeout: float | None = None,
        delay: float | None = None,
        wait_for_selector: str | None = None,
        block_resources: list[str] | None = None,
        headers: dict[str, str] | None = None,
        user_agent: str | None = None,
        callback_url: str | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"url": url, **kwargs}
        if wait_until is not None:
            body["wait_until"] = wait_until
        if timeout is not None:
            body["timeout"] = timeout
        if delay is not None:
            body["delay"] = delay
        if wait_for_selector is not None:
            body["wait_for_selector"] = wait_for_selector
        if block_resources is not None:
            body["block_resources"] = block_resources
        if headers is not None:
            body["headers"] = headers
        if user_agent is not None:
            body["user_agent"] = user_agent
        if callback_url is not None:
            body["callback_url"] = callback_url
        return await self._transport.request("POST", "/webextrator/render", json=body)

File: typescript/src/resources/webextrator.ts
/** WebExtrator web render & extract resources. */

import { Transport } from '../runtime/transport';

class Tasks {
  constructor(private transport: Transport) {}

  async retrieve(opts: {
    id?: string;
    traceId?: string;
    [key: string]: unknown;
  } = {}): Promise<Record<string, unknown>> {
    const { id, traceId, ...rest } = opts;
    const body: Record<string, unknown> = { action: 'retrieve', ...rest };
    if (id !== undefined) body.id = id;
    if (traceId !== undefined) body.trace_id = traceId;
    return this.transport.request('POST', '/webextrator/tasks', { json: body });
  }

  async retrieveBatch(opts: {
    ids?: string[];
    traceIds?: string[];
    offset?: number;
    limit?: number;
    [key: string]: unknown;
  } = {}): Promise<Record<string, unknown>> {
    const { ids, traceIds, offset, limit, ...rest } = opts;
    const body: Record<string, unknown> = { action: 'retrieve_batch', ...rest };
    if (ids !== undefined) body.ids = ids;
    if (traceIds !== undefined) body.trace_ids = traceIds;
    if (offset !== undefined) body.offset = offset;
    if (limit !== undefined) body.limit = limit;
    return this.transport.request('POST', '/webextrator/tasks', { json: body });
  }
}

export class WebExtrator {
  readonly tasks: Tasks;

  constructor(private transport: Transport) {
    this.tasks = new Tasks(transport);
  }

  async extract(opts: {
    url: string;
    expectedType?: 'product' | 'article' | 'general';
    enableLlm?: boolean;
    waitUntil?: 'load' | 'domcontentloaded' | 'networkidle' | 'commit';
    timeout?: number;
    delay?: number;
    waitForSelector?: string;
    blockResources?: string[];
    headers?: Record<string, string>;
    userAgent?: string;
    callbackUrl?: string;
    async?: boolean;
    [key: string]: unknown;
  }): Promise<Record<string, unknown>> {
    const { url, expectedType, enableLlm, waitUntil, timeout, delay, waitForSelector, blockResources, headers, userAgent, callbackUrl, ...rest } = opts;
    const body: Record<string, unknown> = { url, ...rest };
    if (expectedType !== undefined) body.expected_type = expectedType;
    if (enableLlm !== undefined) body.enable_llm = enableLlm;
    if (waitUntil !== undefined) body.wait_until = waitUntil;
    if (timeout !== undefined) body.timeout = timeout;
    if (delay !== undefined) body.delay = delay;
    if (waitForSelector !== undefined) body.wait_for_selector = waitForSelector;
    if (blockResources !== undefined) body.block_resources = blockResources;
    if (headers !== undefined) body.headers = headers;
    if (userAgent !== undefined) body.user_agent = userAgent;
    if (callbackUrl !== undefined) body.callback_url = callbackUrl;
    return this.transport.request('POST', '/webextrator/extract', { json: body });
  }

  async render(opts: {
    url: string;
    waitUntil?: 'load' | 'domcontentloaded' | 'networkidle' | 'commit';
    timeout?: number;
    delay?: number;
    waitForSelector?: string;
    blockResources?: string[];
    headers?: Record<string, string>;
    userAgent?: string;
    callbackUrl?: string;
    async?: boolean;
    [key: string]: unknown;
  }): Promise<Record<string, unknown>> {
    const { url, waitUntil, timeout, delay, waitForSelector, blockResources, headers, userAgent, callbackUrl, ...rest } = opts;
    const body: Record<string, unknown> = { url, ...rest };
    if (waitUntil !== undefined) body.wait_until = waitUntil;
    if (timeout !== undefined) body.timeout = timeout;
    if (delay !== undefined) body.delay = delay;
    if (waitForSelector !== undefined) body.wait_for_selector = waitForSelector;
    if (blockResources !== undefined) body.block_resources = blockResources;
    if (headers !== undefined) body.headers = headers;
    if (userAgent !== undefined) body.user_agent = userAgent;
    if (callbackUrl !== undefined) body.callback_url = callbackUrl;
    return this.transport.request('POST', '/webextrator/render', { json: body });
  }
}
